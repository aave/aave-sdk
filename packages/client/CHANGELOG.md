# @aave/client

## 0.3.1

### Patch Changes

- **fix:** Support max and exact value options in repay function.
- Updated dependencies
  - @aave/graphql@0.3.1

## 0.3.0

### Minor Changes

- **feat:** support for healthFactorPreview query

### Patch Changes

- Updated dependencies
  - @aave/graphql@0.3.0

## 0.2.0

### Minor Changes

- 0e36ce5: **feat:** supports for borrow credit delegation

### Patch Changes

- Updated dependencies [0e36ce5]
  - @aave/graphql@0.2.0

## 0.1.0

### Minor Changes

- fb26904: **feat:** first release

### Patch Changes

- Updated dependencies [fb26904]
  - @aave/graphql@0.1.0
  - @aave/types@0.1.0
