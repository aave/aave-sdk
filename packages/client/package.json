{
  "name": "@aave/client",
  "version": "0.0.0",
  "description": "The official JavaScript client for the Aave API",
  "repository": {
    "directory": "packages/client",
    "type": "git",
    "url": "git://github.com/aave/aave-sdk.git"
  },
  "type": "module",
  "main": "dist/index.cjs",
  "module": "dist/index.js",
  "types": "dist/index.d.ts",
  "exports": {
    ".": {
      "import": "./dist/index.js",
      "require": "./dist/index.cjs"
    },
    "./actions": {
      "import": "./dist/actions/index.js",
      "require": "./dist/actions/index.cjs"
    },
    "./ethers": {
      "import": "./dist/ethers.js",
      "require": "./dist/ethers.cjs"
    },
    "./viem": {
      "import": "./dist/viem.js",
      "require": "./dist/viem.cjs"
    },
    "./test-utils": {
      "import": "./dist/test-utils.js",
      "require": "./dist/test-utils.cjs"
    }
  },
  "typesVersions": {
    "*": {
      "actions": [
        "./dist/actions/index.d.ts"
      ],
      "ethers": [
        "./dist/ethers.d.ts"
      ],
      "viem": [
        "./dist/viem.d.ts"
      ],
      "test-utils": [
        "./dist/test-utils.d.ts"
      ]
    }
  },
  "files": [
    "dist"
  ],
  "sideEffects": false,
  "scripts": {
    "build": "tsup"
  },
  "dependencies": {
    "@aave/env": "workspace:*",
    "@aave/graphql": "workspace:*",
    "@aave/types": "workspace:*",
    "@urql/core": "^5.2.0",
    "graphql": "^16.11.0"
  },
  "devDependencies": {
    "ethers": "^6.14.4",
    "tsup": "^8.5.0",
    "typescript": "^5.6.3",
    "viem": "^2.31.6"
  },
  "peerDependencies": {
    "ethers": "^6.14.4",
    "viem": "^2.31.6"
  },
  "peerDependenciesMeta": {
    "ethers": {
      "optional": true
    },
    "viem": {
      "optional": true
    }
  },
  "license": "MIT"
}
