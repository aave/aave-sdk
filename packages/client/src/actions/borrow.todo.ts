import { describe, expect, it } from 'vitest';

// THIS IS A VERY WIP

describe('Given an Aave Market', () => {
  describe('And a user with a supply position', () => {
    describe('When the user enables e-mode for that market', () => {
      it('Then it should be possible to borrow from the reserve', () => {
        expect(true).toBe(true);
      });
    });
  });
});
