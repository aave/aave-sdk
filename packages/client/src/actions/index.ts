export * from './markets';
export * from './misc';
export * from './permits';
export * from './reserve';
export * from './transactions';
export * from './user';
export * from './vaults';
