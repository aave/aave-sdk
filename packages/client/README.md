# `@aave/client`

The official TypeScript library for interacting with Aave Protocol.

---


The `@aave/client` package contains the core functionality to query markets, execute transactions, and manage user positions across Aave lending markets and vaults. It provides a type-safe interface with built-in error handling and result types.


## Usage

```ts
import { AaveClient, evmAddress, chainId } from '@aave/client';
import { supply, userSupplies } from '@aave/client/actions';
import { sendWith } from '@aave/client/viem';

// Create client
const client = AaveClient.create();

// Query user positions
const positions = await userSupplies(client, {
  markets: [evmAddress('0x87870bca3f3fd6335c3f4ce8392d69350b4fa4e2')],
  user: evmAddress('0x742d35cc6e5c4ce3b69a2a8c7c8e5f7e9a0b1234'),
});

// Execute transactions
const result = await supply(client, {
  market: evmAddress('0x87870bca3f3fd6335c3f4ce8392d69350b4fa4e2'),
  amount: {
    erc20: {
      currency: evmAddress('0xa0b86a33e6441c8c5f0bb9b7e5e1f8bbf5b78b5c'),
      value: '1000'
    }
  },
  supplier: evmAddress('0x742d35cc6e5c4ce3b69a2a8c7c8e5f7e9a0b1234'),
  chainId: chainId(1),
})
  .andThen(sendWith(wallet))
  .andThen(client.waitForTransaction);
```
