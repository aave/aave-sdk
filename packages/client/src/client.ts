import type {
  HasProcessedKnownTransactionRequest,
  StandardData,
} from '@aave/graphql';
import {
  errAsync,
  invariant,
  okAsync,
  ResultAsync,
  type TxHash,
} from '@aave/types';
import {
  type AnyVariables,
  createClient,
  type Exchange,
  fetchExchange,
  type OperationResult,
  type OperationResultSource,
  type TypedDocumentNode,
  type Client as UrqlClient,
} from '@urql/core';
import { hasProcessedKnownTransaction } from './actions';
import { BatchQueryBuilder } from './batch';
import type { ClientConfig } from './config';
import { type Context, configureContext } from './context';
import { TimeoutError, UnexpectedError } from './errors';
import { Logger, LogLevel } from './logger';
import {
  isHasProcessedKnownTransactionRequest,
  type TransactionExecutionResult,
} from './types';
import { delay } from './utils';

function takeValue<T>({
  data,
  error,
}: OperationResult<StandardData<T> | undefined, AnyVariables>): T {
  invariant(data, `Expected a value, got: ${error?.message}`);
  return data.value;
}

export class AaveClient {
  /**
   * @internal
   */
  public readonly urql: UrqlClient;

  private readonly logger: Logger;

  private constructor(private readonly context: Context) {
    this.logger = Logger.named(
      this.constructor.name,
      context.debug ? LogLevel.DEBUG : LogLevel.SILENT,
    );

    this.urql = createClient({
      url: context.environment.backend,
      fetchOptions: {
        credentials: 'omit',
        headers: context.headers,
      },
      exchanges: this.exchanges(),
    });
  }

  /**
   * Create a new instance of the {@link AaveClient}.
   *
   * ```ts
   * const client = AaveClient.create({
   *   environment: production,
   *   origin: 'http://example.com',
   * });
   * ```
   *
   * @param options - The options to configure the client.
   * @returns The new instance of the client.
   */
  static create(options?: ClientConfig): AaveClient {
    return new AaveClient(configureContext(options ?? {}));
  }

  /**
   * Execute a GraphQL query operation.
   *
   * @param document - The GraphQL document to execute.
   * @param variables - The variables to pass to the operation.
   * @returns The result of the operation.
   */
  public query<TValue, TVariables extends AnyVariables>(
    document: TypedDocumentNode<StandardData<TValue>, TVariables>,
    variables: TVariables,
  ): ResultAsync<TValue, UnexpectedError> {
    const query = this.context.fragments.replaceFrom(document);
    return this.resultFrom(this.urql.query(query, variables)).map(takeValue);
  }

  /**
   * Execute a GraphQL mutation operation.
   *
   * @param document - The GraphQL document to execute.
   * @param variables - The variables to pass to the operation.
   * @returns The result of the operation.
   */
  public mutation<TValue, TVariables extends AnyVariables>(
    document: TypedDocumentNode<StandardData<TValue>, TVariables>,
    variables: TVariables,
  ): ResultAsync<TValue, UnexpectedError> {
    return this.resultFrom(this.urql.mutation(document, variables)).map(
      takeValue,
    );
  }

  // TODO update example below
  /**
   * Execute a batch of GraphQL query operations.
   *
   * @alpha This is an alpha API and may be subject to breaking changes.
   *
   * ```ts
   * const result = await client.batch((c) => [
   *   fetchAccount(c, { address: evmAddress('0x1234…') }).map(nonNullable),
   *   fetchBalancesBulk(c, {
   *     includeNative: true,
   *     tokens: [
   *       evmAddress("0x5678…"),
   *       evmAddress("0x9012…"),
   *     ],
   *   }),
   * ]);
   *
   * // const result: Result<
   * //   [
   * //     Account,
   * //     AnyAccountBalance[],
   * //   ],
   * //   UnauthenticatedError | UnexpectedError
   * // >
   * ```
   *
   * @param cb - The callback with the scoped client to execute the actions with.
   * @returns The results of all queries in the same order as they were added.
   */
  batch<T1, T2, E1 extends Error, E2 extends Error>(
    cb: (client: this) => [ResultAsync<T1, E1>, ResultAsync<T2, E2>],
  ): ResultAsync<[T1, T2], E1 | E2>;
  batch<T1, T2, T3, E1 extends Error, E2 extends Error, E3 extends Error>(
    cb: (
      client: this,
    ) => [ResultAsync<T1, E1>, ResultAsync<T2, E2>, ResultAsync<T3, E3>],
  ): ResultAsync<[T1, T2, T3], E1 | E2 | E3>;
  batch<
    T1,
    T2,
    T3,
    T4,
    E1 extends Error,
    E2 extends Error,
    E3 extends Error,
    E4 extends Error,
  >(
    cb: (
      client: this,
    ) => [
      ResultAsync<T1, E1>,
      ResultAsync<T2, E2>,
      ResultAsync<T3, E3>,
      ResultAsync<T4, E4>,
    ],
  ): ResultAsync<[T1, T2, T3, T4], E1 | E2 | E3 | E4>;
  batch<
    T1,
    T2,
    T3,
    T4,
    T5,
    E1 extends Error,
    E2 extends Error,
    E3 extends Error,
    E4 extends Error,
    E5 extends Error,
  >(
    cb: (
      client: this,
    ) => [
      ResultAsync<T1, E1>,
      ResultAsync<T2, E2>,
      ResultAsync<T3, E3>,
      ResultAsync<T4, E4>,
      ResultAsync<T5, E5>,
    ],
  ): ResultAsync<[T1, T2, T3, T4, T5], E1 | E2 | E3 | E4 | E5>;
  batch<
    T1,
    T2,
    T3,
    T4,
    T5,
    T6,
    E1 extends Error,
    E2 extends Error,
    E3 extends Error,
    E4 extends Error,
    E5 extends Error,
    E6 extends Error,
  >(
    cb: (
      client: this,
    ) => [
      ResultAsync<T1, E1>,
      ResultAsync<T2, E2>,
      ResultAsync<T3, E3>,
      ResultAsync<T4, E4>,
      ResultAsync<T5, E5>,
      ResultAsync<T6, E6>,
    ],
  ): ResultAsync<[T1, T2, T3, T4, T5, T6], E1 | E2 | E3 | E4 | E5 | E6>;
  batch<
    T1,
    T2,
    T3,
    T4,
    T5,
    T6,
    T7,
    E1 extends Error,
    E2 extends Error,
    E3 extends Error,
    E4 extends Error,
    E5 extends Error,
    E6 extends Error,
    E7 extends Error,
  >(
    cb: (
      client: this,
    ) => [
      ResultAsync<T1, E1>,
      ResultAsync<T2, E2>,
      ResultAsync<T3, E3>,
      ResultAsync<T4, E4>,
      ResultAsync<T5, E5>,
      ResultAsync<T6, E6>,
      ResultAsync<T7, E7>,
    ],
  ): ResultAsync<
    [T1, T2, T3, T4, T5, T6, T7],
    E1 | E2 | E3 | E4 | E5 | E6 | E7
  >;
  batch<
    T1,
    T2,
    T3,
    T4,
    T5,
    T6,
    T7,
    T8,
    E1 extends Error,
    E2 extends Error,
    E3 extends Error,
    E4 extends Error,
    E5 extends Error,
    E6 extends Error,
    E7 extends Error,
    E8 extends Error,
  >(
    cb: (
      client: this,
    ) => [
      ResultAsync<T1, E1>,
      ResultAsync<T2, E2>,
      ResultAsync<T3, E3>,
      ResultAsync<T4, E4>,
      ResultAsync<T5, E5>,
      ResultAsync<T6, E6>,
      ResultAsync<T7, E7>,
      ResultAsync<T8, E8>,
    ],
  ): ResultAsync<
    [T1, T2, T3, T4, T5, T6, T7, T8],
    E1 | E2 | E3 | E4 | E5 | E6 | E7 | E8
  >;
  batch<
    T1,
    T2,
    T3,
    T4,
    T5,
    T6,
    T7,
    T8,
    T9,
    E1 extends Error,
    E2 extends Error,
    E3 extends Error,
    E4 extends Error,
    E5 extends Error,
    E6 extends Error,
    E7 extends Error,
    E8 extends Error,
    E9 extends Error,
  >(
    cb: (
      client: this,
    ) => [
      ResultAsync<T1, E1>,
      ResultAsync<T2, E2>,
      ResultAsync<T3, E3>,
      ResultAsync<T4, E4>,
      ResultAsync<T5, E5>,
      ResultAsync<T6, E6>,
      ResultAsync<T7, E7>,
      ResultAsync<T8, E8>,
      ResultAsync<T9, E9>,
    ],
  ): ResultAsync<
    [T1, T2, T3, T4, T5, T6, T7, T8, T9],
    E1 | E2 | E3 | E4 | E5 | E6 | E7 | E8 | E9
  >;
  batch<
    T1,
    T2,
    T3,
    T4,
    T5,
    T6,
    T7,
    T8,
    T9,
    T10,
    E1 extends Error,
    E2 extends Error,
    E3 extends Error,
    E4 extends Error,
    E5 extends Error,
    E6 extends Error,
    E7 extends Error,
    E8 extends Error,
    E9 extends Error,
    E10 extends Error,
  >(
    cb: (
      client: this,
    ) => [
      ResultAsync<T1, E1>,
      ResultAsync<T2, E2>,
      ResultAsync<T3, E3>,
      ResultAsync<T4, E4>,
      ResultAsync<T5, E5>,
      ResultAsync<T6, E6>,
      ResultAsync<T7, E7>,
      ResultAsync<T8, E8>,
      ResultAsync<T9, E9>,
      ResultAsync<T10, E10>,
    ],
  ): ResultAsync<
    [T1, T2, T3, T4, T5, T6, T7, T8, T9, T10],
    E1 | E2 | E3 | E4 | E5 | E6 | E7 | E8 | E9 | E10
  >;
  batch<T, E extends Error>(
    cb: (client: this) => ResultAsync<T, E>[],
  ): ResultAsync<T[], E>;
  batch(
    cb: (client: this) => ResultAsync<unknown[], unknown>[],
  ): ResultAsync<unknown[], unknown> {
    const builder = new BatchQueryBuilder();

    const client: this = Object.create(this, {
      query: {
        value: builder.addQuery,
      },
    });

    const combined = ResultAsync.combine(cb(client));
    const [document, variables] = builder.build();

    const query = this.context.fragments.replaceFrom(document);

    return this.resultFrom(this.urql.query(query, variables))
      .andTee(({ data, error }) => {
        invariant(data, `Expected a value, got: ${error?.message}`);
        builder.resolve(data);
      })
      .andThen(() => combined);
  }

  /**
   * @internal
   */
  readonly waitForSupportedTransaction = (
    result: TransactionExecutionResult,
  ): ResultAsync<TxHash, TimeoutError | UnexpectedError> => {
    if (isHasProcessedKnownTransactionRequest(result)) {
      return this.waitForTransaction(result);
    }
    return okAsync(result);
  };

  /**
   * Given the transaction hash of an Aave protocol transaction, wait for the transaction to be
   * processed by the Aave v3 API.
   *
   * Returns a {@link TimeoutError} if the transaction is not processed within the expected timeout period.
   *
   * @param result - The transaction execution result to wait for.
   * @returns The transaction hash or a TimeoutError
   */
  readonly waitForTransaction = (
    result: TransactionExecutionResult,
  ): ResultAsync<TxHash, TimeoutError | UnexpectedError> => {
    invariant(
      isHasProcessedKnownTransactionRequest(result),
      `Received a transaction result for an untracked operation. Make sure you're following the instructions in the docs.`,
    );

    return ResultAsync.fromPromise(
      this.pollTransactionStatus(result),
      (err) => {
        if (err instanceof TimeoutError || err instanceof UnexpectedError) {
          return err;
        }
        return UnexpectedError.from(err);
      },
    );
  };

  protected async pollTransactionStatus(
    request: HasProcessedKnownTransactionRequest,
  ): Promise<TxHash> {
    const startedAt = Date.now();

    while (Date.now() - startedAt < this.context.environment.indexingTimeout) {
      const processed = await hasProcessedKnownTransaction(this, request).match(
        (ok) => ok,
        (err) => {
          throw err;
        },
      );

      if (processed) {
        return request.txHash;
      }

      await delay(this.context.environment.pollingInterval);
    }
    throw TimeoutError.from(
      `Timeout waiting for transaction ${request.txHash} to be processed.`,
    );
  }

  protected exchanges(): Exchange[] {
    return [fetchExchange];
  }

  protected resultFrom<TData, TVariables extends AnyVariables>(
    source: OperationResultSource<OperationResult<TData, TVariables>>,
  ): ResultAsync<OperationResult<TData, TVariables>, UnexpectedError> {
    return ResultAsync.fromPromise(source.toPromise(), (err: unknown) => {
      this.logger.error(err);
      console.log(err);
      return UnexpectedError.from(err);
    }).andThen((result) => {
      if (result.error?.networkError) {
        return errAsync(UnexpectedError.from(result.error.networkError));
      }
      return okAsync(result);
    });
  }
}
