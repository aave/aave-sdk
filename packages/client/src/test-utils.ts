/// <reference path="../../../vite-env.d.ts" />

import { local, staging } from '@aave/env';
import type { AnyVariables, Reserve } from '@aave/graphql';
import { schema } from '@aave/graphql/test-utils';
import {
  assertOk,
  type BigDecimal,
  bigDecimal,
  chainId,
  type EvmAddress,
  evmAddress,
  nonNullable,
  ResultAsync,
} from '@aave/types';
import type { TypedDocumentNode } from '@urql/core';
import { validate } from 'graphql';
import type { ValidationRule } from 'graphql/validation/ValidationContext';
import {
  createPublicClient,
  createWalletClient,
  defineChain,
  http,
  parseEther,
  parseUnits,
  type WalletClient,
} from 'viem';
import { generatePrivateKey, privateKeyToAccount } from 'viem/accounts';
import { expect } from 'vitest';
import { reserve } from './actions';
import { AaveClient } from './client';
import { GraphQLErrorCode, UnexpectedError } from './errors';

export const signer = privateKeyToAccount(import.meta.env.PRIVATE_KEY);
export const environment =
  import.meta.env.ENVIRONMENT === 'local' ? local : staging;

export const ETHEREUM_FORK_ID = chainId(99999999);

export const WETH_ADDRESS = evmAddress(
  '0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2',
);
export const USDC_ADDRESS = evmAddress(
  '0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48',
);
export const DEFAULT_MARKET_ADDRESS = evmAddress(
  '0x87870Bca3F3fD6335C3F4ce8392D69350B4fA4E2',
);

export const ETHEREUM_FORK_RPC_URL =
  'https://virtual.mainnet.rpc.tenderly.co/27ff3c60-0e2c-4d46-8190-f5170dc7da8c';

export const ETHEREUM_FORK_RPC_URL_ADMIN =
  'https://virtual.mainnet.rpc.tenderly.co/95925d93-2ca7-4986-8b4f-e606f6b243bd';

// Re-export for convenience
export { bigDecimal } from '@aave/types';

const ethereumForkChain = defineChain({
  id: ETHEREUM_FORK_ID,
  name: 'Ethereum Fork',
  network: 'ethereum-fork',
  nativeCurrency: { name: 'Ether', symbol: 'ETH', decimals: 18 },
  rpcUrls: { default: { http: [ETHEREUM_FORK_RPC_URL] } },
});

export const client = AaveClient.create({
  environment,
});

export const wallet: WalletClient = createWalletClient({
  account: signer,
  chain: ethereumForkChain,
  transport: http(),
});

export function createNewWallet(): WalletClient {
  const privateKey = generatePrivateKey();
  // Log private key to debug test issues
  console.log(`private key ${privateKey}`);
  const wallet = createWalletClient({
    account: privateKeyToAccount(privateKey),
    chain: ethereumForkChain,
    transport: http(),
  });
  return wallet;
}

// Tenderly RPC type for setBalance
type TSetBalanceRpc = {
  Method: 'tenderly_setBalance';
  Parameters: [addresses: string[], amount: string];
  ReturnType: string;
};

// Tenderly RPC type for set ERC20 balance
type TSetErc20BalanceRpc = {
  Method: 'tenderly_setErc20Balance';
  Parameters: [tokenAddress: string, address: string, amount: string];
  ReturnType: string;
};

export function fundNativeAddress(
  address: EvmAddress,
  amount: BigDecimal = bigDecimal('1.0'), // 1 ETH
): ResultAsync<string, UnexpectedError> {
  // Create client with fork chain - you'll need to replace this with your actual fork chain config
  const publicClient = createPublicClient({
    chain: {
      id: ETHEREUM_FORK_ID,
      name: 'Tenderly Fork',
      network: 'tenderly-fork',
      nativeCurrency: { name: 'Ether', symbol: 'ETH', decimals: 18 },
      rpcUrls: {
        default: { http: [ETHEREUM_FORK_RPC_URL_ADMIN] },
      },
    },
    transport: http(ETHEREUM_FORK_RPC_URL_ADMIN),
  });

  const amountInWei = parseEther(amount);
  const amountHex = `0x${amountInWei.toString(16)}`;

  return ResultAsync.fromPromise(
    publicClient.request<TSetBalanceRpc>({
      method: 'tenderly_setBalance',
      params: [[address], amountHex],
    }),
    (err) => UnexpectedError.from(err),
  );
}

export function fundErc20Address(
  tokenAddress: EvmAddress,
  address: EvmAddress,
  amount: BigDecimal,
  decimals = 18,
): ResultAsync<string, UnexpectedError> {
  const publicClient = createPublicClient({
    chain: {
      id: ETHEREUM_FORK_ID,
      name: 'Tenderly Fork',
      network: 'tenderly-fork',
      nativeCurrency: { name: 'Ether', symbol: 'ETH', decimals: 18 },
      rpcUrls: {
        default: { http: [ETHEREUM_FORK_RPC_URL_ADMIN] },
      },
    },
    transport: http(ETHEREUM_FORK_RPC_URL_ADMIN),
  });

  // Convert amount to the smallest unit (e.g., wei for 18 decimals)
  const amountInSmallestUnit = parseUnits(amount, decimals);
  const amountHex = `0x${amountInSmallestUnit.toString(16)}`;

  return ResultAsync.fromPromise(
    publicClient.request<TSetErc20BalanceRpc>({
      method: 'tenderly_setErc20Balance',
      params: [tokenAddress, address, amountHex],
    }),
    (err) => UnexpectedError.from(err),
  );
}

export async function getReserveInfo(
  tokenAddress: EvmAddress,
  marketAddress = DEFAULT_MARKET_ADDRESS,
): Promise<Reserve> {
  const result = await reserve(client, {
    chainId: ETHEREUM_FORK_ID,
    market: marketAddress,
    underlyingToken: tokenAddress,
  }).map(nonNullable);
  assertOk(result);
  return result.value;
}

const messages: Record<GraphQLErrorCode, string> = {
  [GraphQLErrorCode.UNAUTHENTICATED]:
    "Unauthenticated - Authentication is required to access '<operation>'",
  [GraphQLErrorCode.FORBIDDEN]:
    "Forbidden - You are not authorized to access '<operation>'",
  [GraphQLErrorCode.INTERNAL_SERVER_ERROR]:
    'Internal server error - Please try again later',
  [GraphQLErrorCode.BAD_USER_INPUT]:
    'Bad user input - Please check the input and try again',
  [GraphQLErrorCode.BAD_REQUEST]:
    'Bad request - Please check the request and try again',
};

export function createGraphQLErrorObject(code: GraphQLErrorCode) {
  return {
    message: messages[code],
    locations: [],
    path: [],
    extensions: {
      code: code,
    },
  };
}

export function assertTypedDocumentSatisfies<
  TResult,
  TVariables extends AnyVariables,
>(
  document: TypedDocumentNode<TResult, TVariables>,
  rules: ReadonlyArray<ValidationRule>,
) {
  expect(validate(schema, document, rules)).toEqual([]);
}
