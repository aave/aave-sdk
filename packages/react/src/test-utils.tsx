import { AaveClient } from '@aave/client';
import { environment } from '@aave/client/test-utils';
import { type RenderHookOptions, renderHook } from '@testing-library/react';
// biome-ignore lint/correctness/noUnusedImports: needed for types
import React, { Component, type ReactNode, Suspense } from 'react';
import { AaveContextProvider } from './context';

class TestErrorBoundary extends Component<{ children: ReactNode }> {
  state: { error?: Error } = {};

  static getDerivedStateFromError(error: Error) {
    return { hasError: true, error };
  }

  render() {
    if (this.state.error) {
      return (
        <div data-testid='error-boundary'>
          Error: {this.state.error?.message}
        </div>
      );
    }

    return this.props.children;
  }
}

function createTestWrapper() {
  const client = AaveClient.create({
    environment,
  });

  return function TestWrapper({ children }: { children: ReactNode }) {
    return (
      <AaveContextProvider client={client}>
        <TestErrorBoundary>
          <Suspense fallback={<div data-testid='loading'>Loading...</div>}>
            {children}
          </Suspense>
        </TestErrorBoundary>
      </AaveContextProvider>
    );
  };
}

export type RenderHookWithContextOptions<TProps> = Omit<
  RenderHookOptions<TProps>,
  'wrapper'
> & {
  onError?: (error: Error) => void;
};

/**
 * Unified wrapper around `renderHook` from `@testing-library/react`.
 *
 * All hooks are wrapped with:
 * - AaveContextProvider (required for Aave hooks)
 * - Suspense boundary (with loading fallback)
 * - ErrorBoundary (requires custom error handler)
 *
 * ```ts
 * const { result } = renderHookWithinContext(() => useMyHook(), {
 *   client: mockClient,
 * });
 * ```
 */
export function renderHookWithinContext<TProps, TResult>(
  callback: (props: TProps) => TResult,
  options: RenderHookWithContextOptions<TProps> = {},
) {
  return renderHook(callback, {
    wrapper: createTestWrapper(),
    ...options,
  });
}
