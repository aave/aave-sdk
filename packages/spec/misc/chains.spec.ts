import { assertOk, ChainsFilter } from '@aave/client';
import { chains } from '@aave/client/actions';
import { client } from '@aave/client/test-utils';
import { describe, expect, it } from 'vitest';

describe('Given the Aave Protocol', () => {
  describe('When listing all supported chains', () => {
    it('Then it should return the expected list chains', async () => {
      const result = await chains(client, ChainsFilter.ALL);
      assertOk(result);
      // Sort by chainId to make the snapshot stable
      const sortedChains = result.value.sort((a, b) => a.chainId - b.chainId);
      expect(sortedChains).toMatchSnapshot();
    });
  });
});
