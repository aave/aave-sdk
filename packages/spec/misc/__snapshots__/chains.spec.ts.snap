// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Given the Aave Protocol > When listing all supported chains > Then it should return the expected list chains 1`] = `
[
  {
    "__typename": "Chain",
    "chainId": 1,
    "explorerUrl": "https://etherscan.io",
    "icon": "https://statics.aave.com/ethereum.svg",
    "isTestnet": false,
    "name": "Ethereum",
    "nativeWrappedToken": "0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2",
  },
  {
    "__typename": "Chain",
    "chainId": 10,
    "explorerUrl": "https://optimistic.etherscan.io",
    "icon": "https://statics.aave.com/optimism.svg",
    "isTestnet": false,
    "name": "Optimism",
    "nativeWrappedToken": "0x4200000000000000000000000000000000000006",
  },
  {
    "__typename": "Chain",
    "chainId": 56,
    "explorerUrl": "https://bscscan.com",
    "icon": "https://statics.aave.com/bnbchain.svg",
    "isTestnet": false,
    "name": "BSC",
    "nativeWrappedToken": "0xbb4CdB9CBd36B01bD1cBaEBF2De08d9173bc095c",
  },
  {
    "__typename": "Chain",
    "chainId": 100,
    "explorerUrl": "https://gnosisscan.io",
    "icon": "https://statics.aave.com/Gnosis.svg",
    "isTestnet": false,
    "name": "Gnosis",
    "nativeWrappedToken": "0xe91D153E0b41518A2Ce8Dd3D7944Fa863463a97d",
  },
  {
    "__typename": "Chain",
    "chainId": 137,
    "explorerUrl": "https://polygonscan.com",
    "icon": "https://statics.aave.com/polygon.svg",
    "isTestnet": false,
    "name": "Polygon",
    "nativeWrappedToken": "0x0d500B1d8E8eF31E21C99d1Db9A6444d3ADf1270",
  },
  {
    "__typename": "Chain",
    "chainId": 146,
    "explorerUrl": "https://sonicscan.org",
    "icon": "https://statics.aave.com/sonic.svg",
    "isTestnet": false,
    "name": "Sonic",
    "nativeWrappedToken": "0x039e2fB66102314Ce7b64Ce5Ce3E5183bc94aD38",
  },
  {
    "__typename": "Chain",
    "chainId": 324,
    "explorerUrl": "https://era.zksync.network",
    "icon": "https://statics.aave.com/zksync-era-light.svg",
    "isTestnet": false,
    "name": "zkSync",
    "nativeWrappedToken": "0x5AEa5775959fBC2557Cc8789bC1bf90A239D9a91",
  },
  {
    "__typename": "Chain",
    "chainId": 1088,
    "explorerUrl": "https://andromeda-explorer.metis.io",
    "icon": "https://statics.aave.com/Metis.svg",
    "isTestnet": false,
    "name": "Metis",
    "nativeWrappedToken": "0x75cb093E4D61d2A2e65D8e0BBb01DE8d89b53481",
  },
  {
    "__typename": "Chain",
    "chainId": 1868,
    "explorerUrl": "https://soneium.blockscout.com",
    "icon": "https://statics.aave.com/soneium.svg",
    "isTestnet": false,
    "name": "Soneium",
    "nativeWrappedToken": "0x4200000000000000000000000000000000000006",
  },
  {
    "__typename": "Chain",
    "chainId": 8453,
    "explorerUrl": "https://basescan.org",
    "icon": "https://statics.aave.com/base.svg",
    "isTestnet": false,
    "name": "Base",
    "nativeWrappedToken": "0x4200000000000000000000000000000000000006",
  },
  {
    "__typename": "Chain",
    "chainId": 9745,
    "explorerUrl": "https://plasmascan.to/",
    "icon": "https://statics.aave.com/plasma.svg",
    "isTestnet": false,
    "name": "Plasma",
    "nativeWrappedToken": "0x6100E367285b01F48D07953803A2d8dCA5D19873",
  },
  {
    "__typename": "Chain",
    "chainId": 42161,
    "explorerUrl": "https://arbiscan.io",
    "icon": "https://statics.aave.com/arbitrum.svg",
    "isTestnet": false,
    "name": "Arbitrum",
    "nativeWrappedToken": "0x82aF49447D8a07e3bd95BD0d56f35241523fBab1",
  },
  {
    "__typename": "Chain",
    "chainId": 42220,
    "explorerUrl": "https://celoscan.io",
    "icon": "https://statics.aave.com/celo.svg",
    "isTestnet": false,
    "name": "Celo",
    "nativeWrappedToken": "0xE452E6Ea2dDeB012e20dB73bf5d3863A3Ac8d77a",
  },
  {
    "__typename": "Chain",
    "chainId": 43114,
    "explorerUrl": "https://cchain.explorer.avax.network",
    "icon": "https://statics.aave.com/avalanche.svg",
    "isTestnet": false,
    "name": "Avalanche",
    "nativeWrappedToken": "0xB31f66AA3C1e785363F0875A1B74E27b85FD66c7",
  },
  {
    "__typename": "Chain",
    "chainId": 59144,
    "explorerUrl": "https://lineascan.build",
    "icon": "https://statics.aave.com/linea.svg",
    "isTestnet": false,
    "name": "Linea",
    "nativeWrappedToken": "0xe5D7C2a44FfDDf6b295A15c148167daaAf5Cf34f",
  },
  {
    "__typename": "Chain",
    "chainId": 84532,
    "explorerUrl": "https://sepolia.basescan.org",
    "icon": "https://statics.aave.com/base.svg",
    "isTestnet": true,
    "name": "Base Sepolia",
    "nativeWrappedToken": "0x4200000000000000000000000000000000000006",
  },
  {
    "__typename": "Chain",
    "chainId": 534352,
    "explorerUrl": "https://scrollscan.com",
    "icon": "https://statics.aave.com/scroll-network.svg",
    "isTestnet": false,
    "name": "Scroll",
    "nativeWrappedToken": "0x5300000000000000000000000000000000000004",
  },
  {
    "__typename": "Chain",
    "chainId": 99999999,
    "explorerUrl": "https://dashboard.tenderly.co/explorer/vnet/27ff3c60-0e2c-4d46-8190-f5170dc7da8c",
    "icon": "https://statics.aave.com/ethereum.svg",
    "isTestnet": true,
    "name": "Forked Ethereum",
    "nativeWrappedToken": "0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2",
  },
]
`;
